apisix/plugins/cas-auth.lua

```diff
@@ -0,0 +1,208 @@
+        ", cookie: ", ngx.var.http_cookie, ", request_uri: ", ngx.var.request_uri)
+    set_our_cookie(CAS_REQUEST_URI, ngx.var.request_uri)
+    ngx.redirect(login_uri, ngx.HTTP_MOVED_TEMPORARILY)
+end
+
+local function with_session_id(conf, session_id)
+    -- does the cookie exist in our store?
+    local user = store:get(session_id);
+    ngx.log(ngx.INFO, "ticket=", session_id, ", user=", user)
+    if user == nil then
+        set_our_cookie(COOKIE_NAME, "deleted; Max-Age=0")
+        first_access(conf)
+    else
+        -- refresh the TTL
+        store:set(session_id, user, SESSION_LIFETIME)
+    end
+end
+
+local function set_store_and_cookie(session_id, user)
+    -- place cookie into cookie store
+    local success, err, forcible = store:add(session_id, user, SESSION_LIFETIME)
+    if success then
+        if forcible then
+            ngx.log(ngx.INFO, "CAS cookie store is out of memory")
+        end
+        set_our_cookie(COOKIE_NAME, session_id)
+    else
+        if err == "no memory" then
+            ngx.log(ngx.EMERG, "CAS cookie store is out of memory")
+        elseif err == "exists" then
+            ngx.log(ngx.ERR, "Same CAS ticket validated twice, this should never happen!")
+        end
+    end
+    return success
+end
+
+local function validate(conf, ticket)
+    -- send a request to CAS to validate the ticket
+    local httpc = http.new()
+    local res, err = httpc:request_uri(conf.idp_uri ..
+        "/serviceValidate", { query = { ticket = ticket, service = uri_without_ticket(conf) } })
+
+    if res and res.status == ngx.HTTP_OK and res.body ~= nil then
+        if string.find(res.body, "<cas:authenticationSuccess>") then
+            local m = ngx.re.match(res.body, "<cas:user>(.*?)</cas:user>");
+            if m then
+                return m[1]
+            end
+        else
+            ngx.log(ngx.INFO, "CAS serviceValidate failed: " .. res.body)
+        end
+    else
+        ngx.log(ngx.ERR, "validate ticket failed: res=", res, ", err=", err)
+    end
+    return nil
+end
+
+local function validate_with_cas(conf, ticket)
+    local user = validate(conf, ticket)
+    if user and set_store_and_cookie(ticket, user) then
+        local request_uri = ngx.var["cookie_" .. CAS_REQUEST_URI]
+        set_our_cookie(CAS_REQUEST_URI, "deleted; Max-Age=0")
+        ngx.log(ngx.INFO, "ticket: ", ticket,
+            ", cookie: ", ngx.var.http_cookie, ", request_uri: ", request_uri, ", user=", user)
+        ngx.redirect(request_uri, ngx.HTTP_MOVED_TEMPORARILY)
+    else
+        return ngx.HTTP_UNAUTHORIZED, {message = "invalid ticket"}
+    end
+end
+
+local function logout(conf)
+    local session_id = get_session_id()
+    if session_id == nil then
+        return ngx.HTTP_UNAUTHORIZED
+    end
+
+    ngx.log(ngx.INFO, "logout: ticket=", session_id, ", cookie=", ngx.var.http_cookie)
+    store:delete(session_id)
+    set_our_cookie(COOKIE_NAME, "deleted; Max-Age=0")
+
+    ngx.redirect(conf.idp_uri .. "/logout")
```

Comment:
ngx.redirect works like `ngx.exit`. As we avoid `ngx.exit`, we prefer to do it by ourselves, see
https://github.com/apache/apisix/blob/9129572f6302300ddeee3f2234393445d8ead781/apisix/plugins/redirect.lua#L257